File: algorithms/python/validateStackSequences/validateStackSequences.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/validate-stack-sequences/
# Author: Miao Zhang
# Date:   2021-03-28

class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        for p in pushed:
            stack.append(p)
            while stack and stack[-1] == popped[0]:
                    stack.pop()
                    popped.pop(0)
        return len(stack) == 0

File: algorithms/cpp/validateStackSequences/validateStackSequences.cpp
// Source: https://leetcode.com/problems/validate-stack-sequences/
// Author: Miao Zhang
// Date:   2021-03-28

class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> st;
        auto it = begin(popped);
        for (auto& p: pushed) {
            st.push(p);
            while (!st.empty() && st.top() == *it) {
                st.pop();
                it++;
            }
        }
        return it == end(popped);
    }
};
